src/charm.py

```diff
@@ -17,13 +17,53 @@
 from charms.nginx_ingress_integrator.v0.ingress import IngressRequires
 from charms.prometheus_k8s.v0.prometheus import PrometheusProvider
 from charms.alertmanager_k8s.v0.alertmanager import AlertmanagerConsumer
+from ops.relation import ConsumerBase
+from ops.framework import EventSource, EventBase
+
 
 PROMETHEUS_CONFIG = "/etc/prometheus/prometheus.yml"
 RULES_DIR = "/etc/prometheus/rules"
 
 logger = logging.getLogger(__name__)
 
 
+class IngressConsumer(ConsumerBase):
```

Comment:
Good idea. But why not create a PR against the Ingress charm. Also may be a good idea to factor this out into a separate file `src/ingress.py` and if required any unit tests for it `tests/test_ingress.py`.

Reply:
I'd be in favor of proposing this as a PR for the upstream repo

Reply:
@mthaddon @jnsgruk what do you think about adding this as a library to [charm-k8s-ingress](https://launchpad.net/charm-k8s-ingress)? ([diff ref](https://github.com/canonical/prometheus-operator/pull/97/commits/5ccc81b5736e34e8ea84afa810e6a3d45fbceda0))

Reply:
What are we proposing to add? The `ingress-changed` event? If so, I don't see a strong reason why that couldn't be added to @mthaddon's upstream library.

Reply:
Essentially replace
`self.ingress = IngressRequires(...)`
with
`self.ingress_lib = IngressConsumer(...)`

to be more precise:
1. Add a consumer library - new class that inherits from `ops.relation.ConsumerBase`
2. Delegate all relevant event observation to the new class
3. Expect users to observe a (custom) event coming from the new lib, `self.ingress_lib.ingress_changed`

Whoever currently uses `IngressRequires` directly could still continue and do it.